GitHub actions

peterhashair opened the pull request:
### Description:
All the files that failed will be downloadable in the artifact at the moment. Add a new command for GitHub action. Not sure if that needed `./console tests:sync-ui-screenshots -a github -r matomo-org/matomo 1528756403` just used for my local screen compare and update image.

### Failed UI
![image](https://user-images.githubusercontent.com/3783072/145905312-334f9805-af74-4b7c-961d-01c49e2ae5a7.png)
 


Group all the bash to `.github/scripts/bash/prepare.sh`
This file split PHP XML into smaller fragments `tests/PHPUnit/formatXML.php`

### MYSQLI adapter setup
`mysqli.allow_local_infile = On`

### Submodule implementation: 
To set up Submodule on GitHub need to update plugin_tests.yml. Once approved, github-action should be 4.x-dev
```yaml
    env:
      TARGET_BRANCH: github-aciton
      PLUGIN: AnonymousPiwikUsageMeasurement

```

- AnonymousPiwikUsageMeasurement https://github.com/matomo-org/plugin-AnonymousPiwikUsageMeasurement/pull/67 √
- Bandwidth https://github.com/matomo-org/plugin-Bandwidth/pull/40 √
- SecurityInfo https://github.com/matomo-org/plugin-SecurityInfo/pull/57 
- MarketingCampaignsReporting https://github.com/matomo-org/plugin-MarketingCampaignsReporting/pull/89
- CustomerAlerts https://github.com/matomo-org/plugin-CustomAlerts/pull/104
- CustomVariables https://github.com/matomo-org/plugin-CustomVariables/pull/29
- DeviceDetectorCache https://github.com/matomo-org/plugin-DeviceDetectorCache/pull/33
- LogViewer https://github.com/matomo-org/plugin-LogViewer/pull/40
- VisitorGenerator https://github.com/matomo-org/plugin-VisitorGenerator/pull/86
- LoginLdap https://github.com/matomo-org/plugin-LoginLdap/pull/256
- TasksTimetable https://github.com/matomo-org/plugin-TasksTimetable/pull/29
- QueuedTracking https://github.com/matomo-org/plugin-QueuedTracking/pull/166
- tag-manager https://github.com/matomo-org/tag-manager/pull/386
- TreemapVisualization https://github.com/matomo-org/plugin-TreemapVisualization/pull/25
- TrackingSpamPrevention https://github.com/matomo-org/plugin-TrackingSpamPrevention/pull/40



### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
update folder permission

update folder permission

Commit:
Update matomo-tests.yml

check user group

Commit:
update chown

update chown

Commit:
Update matomo-tests.yml

updates

Commit:
Update matomo-tests.yml

update permission

Commit:
Update matomo-tests.yml

show folder permission after failed

Commit:
Update matomo-tests.yml

update user group

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

add docker to group

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

update user group

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

Commit:
revert to 7.2

revert to 7.2

Commit:
revert to php 8.0

revert to php 8.0

Commit:
Update matomo-tests.yml

update

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

Commit:
test

test

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

Commit:
Update github.conf

Commit:
Update matomo-tests.yml

Commit:
updat

updat

Commit:
Update matomo-tests.yml

Commit:
update fpm

update fpm

Commit:
Update matomo-tests.yml

Commit:
update

update

Commit:
display 502 error

display 502 error

Commit:
Update matomo-tests.yml

Commit:
Update www.conf

Commit:
Update matomo-tests.yml

Commit:
Update www.conf

Commit:
add error log

add error log

Commit:
Update matomo-tests.yml

Commit:
update runner

update runner

Commit:
check fpm error log

check fpm error log

Commit:
update php-fpm.conf

update php-fpm.conf

Commit:
show log

show log

Commit:
Update matomo-tests.yml

test

Commit:
update socket

update socket

Commit:
update test

update test

Commit:
update php

update php

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

Commit:
remove logo update

remove logo update

Commit:
Update matomo-tests.yml

update chrome

Commit:
update 2 screenshot

Commit:
update 2 screenshot

Commit:
update screen shots

update screen shots

Commit:
update screen shots

update screen shots

Commit:
update screenshoots

update screenshoots

Commit:
update screenshot

update screenshot

Commit:
remove browser test on github action

remove browser test on github action

Commit:
update some screenshot shots

update some screenshot shots

Commit:
check where is chome

check where is chome

Commit:
update user agent

update user agent

Commit:
test chrome core

test chrome core

Commit:
update test user agent

update test user agent

Commit:
update browser support

update browser support

Commit:
update screen shot

update screen shot

peterhashair commented:
I think this is kind of working. coupe random fails 
![image](https://user-images.githubusercontent.com/3783072/141243515-8a5a3e7d-213b-4b7a-9630-dc1f44f8704b.png)

Commit:
update UI tests

update UI tests

Commit:
test wait for time

test wait for time

Commit:
update screen shot

update screen shot

Commit:
update UI test

update UI test

peterhashair commented:
I guess this is ready for review, just 2 submodule image needs an update.
update UIs are all sparkline-related. Somehow the sparkline got 100px diff, although it looks the same.

I am not too sure about those. @sgiehl do you know why those `$` not find is failing
[https://github.com/matomo-org/matomo/runs/4175351607?check_suite_focus=true](https://github.com/matomo-org/matomo/runs/4175351607?check_suite_focus=true)  

`
  1) Overlay
       should load correctly:
          Evaluation failed: ReferenceError: $ is not defined
           at __puppeteer_evaluation_script__:2:17
            Url to reproduce: http://localhost/tests/PHPUnit/proxy/?module=Overlay&period=year&date=today&idSite=3#?l=http$3A$2F$2F127.0.0.1$2Ftests$2Fresources$2Foverlay-test-site-real$2F
            Screenshot of failure: /home/runner/work/matomo/matomo/tests/UI/processed-ui-screenshots/should_load_correctly_failure.png
  `

sgiehl commented:
Haven't had a closer look yet, but regarding

> I am not too sure about those. @sgiehl do you know why those $ not find is failing
https://github.com/matomo-org/matomo/runs/4175351607?check_suite_focus=true

Those tests are trying to open a page on `http://127.0.0.1/tests/resources/overlay-test-site-real/`, which should be served by the configured webserver. Not sure if your implementation actually serves something using the IP or maybe creating that site fails for some reason here: https://github.com/matomo-org/matomo/blob/2404b24c3a52f04475b63f300269a129399882ec/tests/PHPUnit/Fixtures/UITestFixture.php#L282

tsteur commented:
I haven't tested it but something that will need to be looked at is to make this work for our many plugins. With travis, a plugin only adds a `.travis.yml` and it works see eg https://github.com/matomo-org/plugin-QueuedTracking/blob/4.x-dev/.travis.yml . This `travis.yml`  doesn't have much logic as otherwise developers might often need to change the plugin travis.yml and they wouldn't know what needs changing.

I don't know what this all means for the code in here. Effectively, a plugin should only have to create a workflow yml with ideally little logic to make the tests work.

I would assume that some of these scripts in the matomo-tests.yml will need to be moved to external scripts 
![image](https://user-images.githubusercontent.com/273120/141370080-053328cd-b9e3-4a4a-8a3f-d66d9f9e1cec.png)

Also I'm assuming that for plugins the directory structure will be different 
![image](https://user-images.githubusercontent.com/273120/141370208-f6ee38d3-1f2c-4377-b856-a98027d96707.png)

Not sure if we want to cover this separately in a separate PR or what is best/easiest? 

I'm mostly mentioning in case this means there will be bigger refactorings needed in the way it works.

peterhashair commented:
@tsteur right, I was wondering why in bash it has so many variables set for plugin names, now I get it. Maybe I should convert all the run command into a bash file take env variables like it was. Then we can reuse it in the plugins, but I guess all the plugins have `.travis.yml` has to convert to GitHub actions as well? what would be the best approach?

tsteur commented:
> . Then we can reuse it in the plugins, but I guess all the plugins have .travis.yml has to convert to GitHub actions as well? 

That will be needed indeed. 

We'll also need to set up the GitHub secrets for each plugin as currently the secrets are configured in Travis.

We'll also need to adjust various docs like https://developer.matomo.org/guides/tests-travis and https://developer.matomo.org/guides/tests-travis-extended

> Or we can leave Travis for plugins for now. Only Core on Github Action. what would be the best approach?

I would say they should all use the same system. That will simplify a lot of things.

peterhashair commented:
@tsteur that makes sense. I was wondering if we make a docker image for this. Because converting to bash and run it in Github Action like the Travis it becomes a unique format, at some stage we move again, it needs to be rewritten as well. But if we build a docker image, it can be portable to any test engine, even locally as long as docker exist.

tsteur commented:
@matomo-org/core-reviewers are there any thoughts on using docker? Any benefits or downsides there?

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/Dashboard/tests/UI/expected-screenshots/DashboardManager_create_new.png
#	plugins/Dashboard/tests/UI/expected-screenshots/DashboardManager_removed.png
#	plugins/Dashboard/tests/UI/expected-screenshots/Dashboard_removed.png
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_individual_goal.png
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_individual_goal_updated.png
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_overview_row_evolution_reloaded.png
#	plugins/Installation/tests/UI/expected-screenshots/Installation_superuser_de.png
#	plugins/Installation/tests/UI/expected-screenshots/Installation_system_check.png
#	tests/UI/expected-screenshots/EmptySite_emptySiteDashboard_ignored.png
#	tests/UI/expected-screenshots/Theme_home.png
#	tests/UI/expected-screenshots/UIIntegrationTest_admin_home.png
#	tests/UI/expected-screenshots/UIIntegrationTest_api_error.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard1.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard2.png
#	tests/UI/specs/UIIntegration_spec.js

Commit:
update bash

update bash

Commit:
update bash on github

update bash on github

Commit:
Update matomo-tests.yml

update path

Commit:
Update matomo-tests.yml

update path

Commit:
Update prepare.sh

update file to be executable

Commit:
update bash color

update bash color

Commit:
update tests

update tests

Commit:
remove bash accept php-fpm

remove bash accept php-fpm

Commit:
update tests

update tests

Commit:
Update FrontControllerTest.php

remove github specialist tests

Commit:
update screenshots

update screenshots

Commit:
Merge branch '4.x-dev' into github-aciton

Commit:
update tests

update tests

Commit:
Update prepare.sh

update tests

Commit:
update js tests

update js tests

Commit:
Update prepare.sh

update bash

Commit:
Update prepare.sh

update tests

Commit:
Update prepare.sh

update script

Commit:
Update prepare.sh

update typo

Commit:
update screen

update screen

peterhashair commented:
@tsteur convert it to bash, still fixing some ui failed mismatch, should I start covert some of the plugin travis.

tsteur commented:
@peterhashair yeah it be great to test a plugin and checking that it works running tests on github actions just by putting in a minor workflow.yml. Not sure if we need to merge this one before to make this work? 

Before doing this, maybe someone should also review this and we'd want tests to pass already here ideally?

peterhashair commented:
@tsteur yes I will fix all the failed tests, I had a preview most of them are tiny fonts different.

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites.png
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites_changed_sort_order.png
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites_page_1.png
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites_search.png
#	tests/UI/expected-screenshots/EmptySite_emptySiteDashboard_ignored.png
#	tests/UI/expected-screenshots/Theme_home.png
#	tests/UI/expected-screenshots/UIIntegrationTest_admin_home.png
#	tests/UI/expected-screenshots/UIIntegrationTest_api_error.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard1.png
#	tests/UI/expected-screenshots/UIIntegrationTest_widgetize_allwebsites.png

Commit:
Update prepare.sh

setup fonts

Commit:
Update prepare.sh

update bash

Commit:
update screenshot

update screenshot

peterhashair commented:
fix almost every UI Test just a couple of common ones remain. 
![image](https://user-images.githubusercontent.com/3783072/144348469-90bb28d9-2205-43ef-9a3e-fe89fa310501.png)

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_overview.png

Commit:
add bash for plugin

add bash for plugin

Commit:
Update AnonymousPiwikUsageMeasurement

add github action for plugin

Commit:
update scrpits

Commit:
Update AnonymousPiwikUsageMeasurement

update commit

Commit:
Update AnonymousPiwikUsageMeasurement

Commit:
Update AnonymousPiwikUsageMeasurement

update

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/Dashboard/tests/UI/expected-screenshots/DashboardManager_removed.png
#	plugins/Dashboard/tests/UI/expected-screenshots/Dashboard_removed.png
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_individual_goal.png
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_individual_goal_updated.png
#	plugins/PagePerformance/tests/UI/expected-screenshots/PagePerformance_load.png
#	tests/UI/expected-screenshots/Theme_home.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard1.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard2.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard3.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard4.png
#	tests/UI/expected-screenshots/UIIntegrationTest_visitors_engagement.png
#	tests/UI/expected-screenshots/UIIntegrationTest_visitors_locations_provider.png
#	tests/UI/expected-screenshots/ViewDataTableTest_subtables_loaded.png

Commit:
Update pluginTests.sh

update to executable

Commit:
update tests

Commit:
update tests

Commit:
Update pluginTests.sh

update path

Commit:
Update pluginTests.sh

update path

Commit:
Update pluginTests.sh

update phpunit.xml path

Commit:
enable UI tests

Commit:
enable UI tests

Commit:
remove testsRun UI

remove testsRun UI

Commit:
update tests

Commit:
update test

Commit:
update submodule checkout

Commit:
update submodule checkout

Commit:
update screenshots

update screenshots

Commit:
update plugin target

update plugin target

Commit:
update plugins

update plugins

Commit:
update screenshots

update screenshots

Commit:
Update SyncScreenshots.php

update test to live

Commit:
Update UsersManager_permissions_bulk_access_set.png

peterhashair commented:
@tsteur convert coupe plugins, it seems working, probably needs someone to review.

sgiehl commented:
@peterhashair As I just updated my tests for device detector: The php install action meanwhile supports PHP 8.1. So in theory you could change that here, so it will already run tests on PHP 8.1

sgiehl commented:
@peterhashair On travis we had builds running on PHP7.2 and PHP8, to cover minimum and maximum supported version. There I would have updated the PHP8 builds only. I didn't yet have a detailed look at this PR, so not sure which builds are running on which PHP version...

Commit:
Update DeviceDetectorCache

update device detector

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

# Conflicts:
#	plugins/CustomVariables
#	plugins/PagePerformance/tests/UI/PagePerformance_spec.js

Commit:
update plugin to github-action

Commit:
Create upload_artifacts.sh

add upload bash

Commit:
Update AnonymousPiwikUsageMeasurement

move back to 4.x

Commit:
move back to 4.x

move back to 4.x

peterhashair commented:
@sgiehl do you want to start reviewing this, it's kind of working, and all the submodules are passing the test except one screenshot I haven't fixed. Also got an update for ui-test-view https://github.com/matomo-org/ui-tests-viewer/pull/37/files.

Commit:
update upload scripts

update upload scripts

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

# Conflicts:
#	plugins/Dashboard/tests/UI/expected-screenshots/DashboardManager_create_new.png
#	plugins/Dashboard/tests/UI/expected-screenshots/Dashboard_reset.png
#	plugins/TestRunner/Commands/SyncScreenshots.php
#	tests/UI/expected-screenshots/EmptySite_emptySiteDashboard_ignored.png

Commit:
update bash to executeble

update bash to executeble

Commit:
print error

print error

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,235 @@
+      fail-fast: false
+      matrix:
+        tests: [ 'UITests' ]
+        parts: [ 0,1,2,3,4,5,6,7 ]
```
Having too many parts might actually even make it slower. The database for our fixtures (e.g. UITestFixture) is reused when it was already set up by a previous test, that for sure can't work for different runners.

sgiehl commented:
@peterhashair Sure I can start reviewing that one. Do you want me to have a look at something specific or is this one fully finished? If so please ensure you had a look yourself, that only required changes are included and that our new test suites are testing the same as on travis (e.g. same suites on same PHP versions). I don't want to waste time looking at unfinished PRs.

Commit:
Update matomo-tests.yml

update php version

Commit:
stop split

stop split

Commit:
Update matomo-tests.yml

try 7.2

Commit:
Update matomo-tests.yml

update to 8.0

Commit:
Update matomo-tests.yml

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

peterhashair commented:
@sgiehl just update the PHP version Should be the same as the Travis now. The upload screenshot (UI compare) won't work, probably required the ui-tests-viewer deploy. Still needs to clean the Travis folder and etc once confirm that's working.
PHP version listed. 
- All the Integration, unit and system and core tests run on 7.2 and 8.0,  
- javascript test run on 7.2
-  UI test run on 7.2.

Commit:
update text

update text

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

# Conflicts:
#	plugins/Installation/tests/UI/expected-screenshots/Installation_system_check.png

Commit:
update test

update test

Commit:
Update DependencyTest.php

update test

Commit:
update tests

update tests

Commit:
Update DependencyTest.php

update test for 7.2

Commit:
Update DependencyTest.php

Commit:
skip 2 test on 7.2 and try ui on 8.0

skip 2 test on 7.2 and try ui on 8.0

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,238 @@
+      fail-fast: false
+      matrix:
+        tests: [ 'UITests' ]
+        php: [ '8.0' ]
```
UI Tests should always run on the minimal required PHP version (7.2. at the moment). That way we can ensure not to use any PHP code features that aren't supported by that version. Running the Integration & System tests on the minimal version isn't enough though, as they cover a lot less than the UI tests

sgiehl commented on core/SupportedBrowser.php:
```diff
@@ -60,7 +61,7 @@ private static function browserNotSupported($shortName, $version)
     }
 
-    private static function throwException($client)
+    private static function throwException($client, $userAgent)
```
Can't see any benefit in that change, as `$userAgent` isn't used...

sgiehl commented on plugins/CoreAdminHome/CustomLogo.php:
```diff
@@ -16,6 +16,7 @@
 use Piwik\Piwik;
 use Piwik\Plugin\Manager;
 use Piwik\SettingsPiwik;
+use function DI\get;
```
```suggestion
```

sgiehl commented on plugins/CoreAdminHome/tests/UI/TrackingFailures_spec.js:
```diff
@@ -13,6 +13,10 @@ describe("TrackingFailures", function () {
     var manageUrl = '?module=CoreAdminHome&action=trackingFailures&idSite=1&period=day&date=today';
     var widgetUrl = '?module=Widgetize&action=iframe&moduleToWidgetize=CoreAdminHome&actionToWidgetize=getTrackingFailures&idSite=1&period=day&date=today&widget=1';
 
+  before(async function(){
+    await page.setUserAgent('Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36');
+  });
```
Just out of curiosity: Was that needed to fix any problem?

peterhashair replied:
@sgiehl that was one of the screenshot, return browser not supported in the error message.

sgiehl commented on tests/PHPUnit/Integration/HttpTest.php:
```diff
@@ -76,6 +76,10 @@ public function testCustomByteRange($method)
         }
 
+        if (($method == 'socket' || $method == 'curl') && getenv('GITHUB')) {
+            $this->markTestSkipped("GITHUB using php -C don't have socket");
```
Are you still using `php -C` ?

sgiehl commented on tests/PHPUnit/Integration/HttpTest.php:
```diff
@@ -106,6 +110,10 @@ public function testHEADOperation($method)
         }
 
+        if ($method == 'socket' && getenv('GITHUB')) {
+            $this->markTestSkipped("GITHUB using php -C don't have socket");
```
Are you still using `php -C` ?

sgiehl commented on .github/scripts/bash/plugins/pluginTests.sh:
```diff
@@ -0,0 +1,84 @@
+            #       older Matomo's will end up w/ a tests:run-ui command that doesn't support it.
+            # a better fix would be to decouple the Matomo testing framework from Matomo in a way that allowed us to
+            # change code for all versions of Matomo as well as selectively for individual Matomo versions.
+            # git checkout 4.x-dev ../../plugins/TestRunner/Commands/TestsRunUI.php
```
That comment can be removed I guess, as that won't be needed for a github action anymore, especially as the `extra-options` flag is supported long enough now.

sgiehl commented on .github/scripts/bash/prepare.sh:
```diff
@@ -0,0 +1,136 @@
+  echo -e "${GREEN}Setup fonts${SET}"
+  git clone --recursive https://github.com/google/woff2.git ../travis_woff2
+  cd ../travis_woff2
+  make clean all
```
Actually I don't think woff2 is required for UI tests. The woff2 file is included in the repo. So that lib is only needed to check if the files is up to date, which is done in a normal test iirc.

sgiehl commented on .github/scripts/bash/prepare.sh:
```diff
@@ -0,0 +1,136 @@
+cp .github/artifacts/config.ini.github.ui.php config/config.ini.php
+cp .github/artifacts/config.dist.js ./tests/UI/config.js
+cp ./tests/PHPUnit/phpunit.xml.dist ./tests/PHPUnit/phpunit.xml
+mkdir -p ./tmp/assets
+mkdir -p ./tmp/cache
+mkdir -p ./tmp/cache/tracker
+mkdir -p ./tmp/latest
+mkdir -p ./tmp/logs
+mkdir -p ./tmp/sessions
+mkdir -p ./tmp/templates_c
+mkdir -p ./tmp/templates_c/d2
+mkdir -p ./tmp/templates_c/2f
+mkdir -p ./tmp/nonexistant
+mkdir -p ./tmp/tcpdf
+mkdir -p ./tmp/climulti
+mkdir -p /tmp
```
If that is always done, why is that already done for UI and JS tests as well a couple of lines above?

sgiehl commented on .github/scripts/bash/prepare.sh:
```diff
@@ -0,0 +1,136 @@
+fi
+
+echo -e "${GREEN}set folder Permission${SET}"
+cp .github/artifacts/config.ini.github.ui.php config/config.ini.php
```
The config that was set for UI and JS tests before, will here be overwritten. Maybe one config for all is enough then? 🤔

sgiehl commented on .github/scripts/bash/prepare.sh:
```diff
@@ -0,0 +1,136 @@
+  mkdir -p ./tests/UI/processed-ui-screenshots
+  chmod a+rw ./tests/lib/geoip-files || true
+  chmod a+rw ./plugins/*/tests/System/processed || true
+  chmod a+rw ./plugins/*/tests/Integration/processed || true
```
That seems also be done earlier already

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,238 @@
+        env:
+          TRAVIS: true
+          GITHUB: true
+          PHP_VERSION: ${{ matrix.php }}
```
Can't this be simplified to:
```suggestion
      - name: Test running
        run: |
          ./vendor/phpunit/phpunit/phpunit --configuration ./tests/PHPUnit/phpunit.xml --testsuite ${{ matrix.tests }}
        env:
          TRAVIS: true
          GITHUB: true
          PHP_VERSION: ${{ matrix.php }}
```
Or do the additional env flags touch the unit tests in any way?

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,238 @@
+        image: redis:5.0
+        ports:
+          - 6379:6379
+        options: --entrypoint redis-server
```
Don't think redis is required for JS tests

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,238 @@
+        ports:
+          - 3306:3306
+        env:
+          MYSQL_DATABASE: piwik_tests
```
We should directly use `matomo` instead of `piwik` where possible
```suggestion
          MYSQL_DATABASE: matomo_tests
```

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,238 @@
+      - name: Setup Node
+        uses: actions/setup-node@v2
+        with:
+          node-version: '12'
```
As we are suggestion to use `Node 16` for developing with Vue, we maybe should try using a newer version in our builds as well (if that doesn't break anything or needs extra effort)

sgiehl commented on tests/PHPUnit/Framework/Fixture.php:
```diff
@@ -615,6 +615,8 @@ public static function getRootUrl()
             $piwikUrl .= $piwikUri;
         }
 
+        $piwikUrl = getenv('GITHUB') ? "http://localhost/" : $piwikUrl;
+
```
Any reason why the action needs a special handling here?

sgiehl commented on tests/PHPUnit/Integration/DependencyTest.php:
```diff
@@ -47,6 +47,10 @@ public function test_getMissingDependencies_EmptyVersion_ShouldBeIgnored()
+        //todo this is need to be fix skip for now
+        if (getenv('PHP_VERSION') == '7.2') {
+            $this->markTestSkipped('version number not exist');
+        }
```
Why shouldn't those tests don't work on 7.2 anymore?

peterhashair replied:
@sgiehl is the PHP version on names format a little different cause problem for this test `7.2.34-28+ubuntu20.04.1+deb.sury.org+1` maybe need a little name format on that test

sgiehl commented on tests/PHPUnit/Integration/SqlTest.php:
```diff
@@ -36,6 +36,10 @@ public function tearDown(): void
     {
+        if(getenv('GITHUB'))
+        {
+            $this->markTestSkipped('Mariadb does\'t work for some reason');
```
`optimize table` should work on MariaDB, so might be good to investigate that.

peterhashair replied:
@sgiehl It seems like OPTIMIZE TABLE works for InnoDB (before MariaDB 10.1.1). https://mariadb.com/kb/en/optimize-table/. I think this one failed. `CREATE TABLE table3 (c INT) ENGINE=InnoDB`.

sgiehl commented on tests/PHPUnit/Integration/TravisEnvironmentTest.php:
```diff
@@ -18,6 +18,10 @@ class TravisEnvironmentTest extends IntegrationTestCase
+        if (getenv('GITHUB')) {
+            $this->markTestSkipped("Github action does't need travis env test");
+        }
+
```
That shouldn't be needed. If the `MYSQL_ADAPTER` env variable isn't set, the test should simply `return` anyways.

sgiehl commented on .github/workflows/matomo-tests.yml:
```diff
@@ -0,0 +1,238 @@
+      fail-fast: false
+      matrix:
+        tests: [ 'UnitTests', 'SystemTestsPlugins' ,'SystemTestsCore' , 'IntegrationTestsCore', 'IntegrationTestsPlugins' ]
+        php: [ '7.2','8.0' ]
```
Before the tests on `8.0` where running the `AllTests` Suite. That made a difference as there might be some tests cases that aren't executed in the other units, due to various reasons (like incorrect folder names,...). So guess we would somehow need to check if all tests are executed correctly at the moment.
In addition the all tests suite was actually running with another database adapter. Would be good to continue running tests on PDO and MYSQLI, as there are differences in the values returned from database.

sgiehl commented on .github/scripts/bash/plugins/pluginTests.sh:
```diff
@@ -0,0 +1,84 @@
+        fi
+    fi
+else
+    if [ "$COVERAGE" = "Unit" ]
+    then
+        echo "Executing tests in test suite UnitTests..."
+        ./vendor/phpunit/phpunit/phpunit --configuration ./tests/PHPUnit/phpunit.xml --testsuite UnitTests --colors --coverage-clover $TRAVIS_BUILD_DIR/build/logs/clover-unit.xml $PHPUNIT_EXTRA_OPTIONS || true
+    elif [ "$COVERAGE" = "Integration" ]
+    then
+        echo "Executing tests in test suite IntegrationTests..."
+        ./vendor/phpunit/phpunit/phpunit --configuration ./tests/PHPUnit/phpunit.xml --testsuite IntegrationTests --colors --coverage-clover $TRAVIS_BUILD_DIR/build/logs/clover-integration.xml $PHPUNIT_EXTRA_OPTIONS || true
+    fi;
```
I would say we can remove all that code coverage stuff. It wasn't used in a long time and if we ever consider using that again, I think it's better to rethink that completely.

sgiehl commented on tests/PHPUnit/formatXML.php:
```diff
@@ -0,0 +1,51 @@
+<?php
```
Is that test splitter still in use? Can see it is used somewhere, but there is actually only one test suite. Guess for the beginning it fine to have similar test splitting as on travis. We can later always look into splitting them up some more...

peterhashair replied:
@sgiehl I guess I can remove this file. it was used in the test, but I realize the max concurrency GitHub action allow is 20. And the exceed part has to wait for the available proccess, it actually makes the whole process slower, when we excceed 20 matrix

sgiehl reviewed (changes requested):
Had a look through the changes and left some comments. In addition:

* You have added the file `plugins/CustomJsTracker/tests/resources/file-made-non-writable.js`, which should be unneeded, as it's a file created during tests. Maybe we could add it to `.gitignore` instead.

Commit:
update some of the test

update some of the test

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

# Conflicts:
#	plugins/Goals/tests/UI/expected-screenshots/GoalsPages_overview.png

Commit:
Update pluginTests.sh

remove unit in plug

Commit:
Update matomo-tests.yml

add mysql adapter

peterhashair commented:
I guess we missing this part in here



When the build fails on Travis
System PHP tests in Matomo typically execute an API method and compare the entire XML output of the API method with an expected XML output.

If you are making changes to Matomo then the result of such an API method may change and break the build. This is an opportunity to review your code and as a Matomo developer you should ensure that any change in the output is actually expected.

If they are not expected, determine the cause of the change and fix it in a new commit. If the changes are expected, then you should update the expected system files accordingly. To compare and update the expected system files, follow these steps:

Find out the Travis build number by opening the Travis run for your pull request. The build number is typically a 5 or 6 digit number and has a leading hash character. For example when the build is #45511, then 45511 is the build number.
Execute this command and replace {buildnumber} with the actual build number. ./console development:sync-system-test-processed {buildnumber}.
To update the expected files directly append the option --expected. You then need to make sure before committing and pushing these changes that every change is actually expected.
Or if you only want to update some files or if you don't use a visual tool for git then you can execute the command without the expected option in which case the system files are updated in the processed directory. For example tests/PHPUnit/System/processed and plugins/Goals/tests/System/processed. If you are using PHPStorm you can then select both the processed and expected directory and then right click -> Compare Directories. This allows you to review every change of added, changed and removed files and let's you update each expected file individually.
Then git add and git commit and git push the changes to trigger another build run.
If some tests are still failing you may need to repeat this process as sometimes you might forget to update some.

Commit:
Update config.dist.js

update ip to localhost

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

Commit:
Update Overlay_spec.js

update to localhost from ip

sgiehl commented:
@peterhashair the system test builds used to upload their artifacts to our build server as well.

Commit:
Update matomo-tests.yml

downgrade mariadb

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

# Conflicts:
#	plugins/Dashboard/tests/UI/expected-screenshots/DashboardManager_removed.png
#	tests/UI/expected-screenshots/Comparison_visitors_overview.png
#	tests/UI/expected-screenshots/Comparison_visitors_overview_segment_removed.png
#	tests/UI/expected-screenshots/Comparison_visitors_overview_switched.png
#	tests/UI/expected-screenshots/Comparison_visits_overview_widget.png
#	tests/UI/expected-screenshots/SupportedBrowser_page_loads_when_browser_supported.png
#	tests/UI/expected-screenshots/Theme_home.png

Commit:
Update pluginTests.sh

remove some comments

Commit:
Update matomo-tests.yml

switch to mysql

Commit:
Update matomo-tests.yml

Commit:
Update matomo-tests.yml

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

Commit:
Update matomo-tests.yml

update adapter

Commit:
Update SqlTest.php

fix Optimize test check

Commit:
Update SqlTest.php

Commit:
Update HttpTest.php

remove getenv

Commit:
Update SqlTest.php

update update test

peterhashair commented:
@sgiehl modified this a little bit, the Mariadb version using the in github action is higher than 10.1.1. `DB::isOptimizeInnoDBSupported()` is returning true. add the bottom function for the test.
```php
      if (DB::isOptimizeInnoDBSupported()) {
            $this->assertTrue(Db::optimizeTables(array('table3', 'table4')) !== false);
        } else {
            $this->assertTrue(Db::optimizeTables(array('table3', 'table4')) === false);
        }
```

Commit:
Update HttpTest.php

update test

Commit:
update test and database name

update test and database name

Commit:
Update .github/workflows/matomo-tests.yml

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Update matomo-tests.yml

update node version

Commit:
Merge branch 'github-aciton' of github.com:matomo-org/matomo into github-aciton

Commit:
update a test

update a test

Commit:
Update DependencyTest.php

update test

Commit:
Update Fixture.php

Commit:
Update matomo-tests.yml

revert node to 12

Commit:
Update prepare.sh

add copy fonts

Commit:
Update matomo-tests.yml

see screenshot

Commit:
Update matomo-tests.yml

update to node 16

Commit:
update tests

update tests

Commit:
update tests

update tests

Commit:
update tests

update tests

Commit:
update tests

update tests

Commit:
Update DependencyTest.php

update tests

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

# Conflicts:
#	plugins/Diagnostics/tests/UI/expected-screenshots/Diagnostics_page.png

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

Commit:
remove socket

remove socket

Commit:
Update matomo-tests.yml

update tests

Commit:
Update HttpTest.php

update tests

Commit:
split php 7.2 and 8.0 tests

split php 7.2 and 8.0 tests

Commit:
Update LogAggregatorTest.php

skip one tests

Commit:
Update matomo-tests.yml

update tests

Commit:
update matrix

update matrix

Commit:
Update matomo-tests.yml

update tests

Commit:
Update matomo-tests.yml

update tests

Commit:
Update prepare.sh

update file config

Commit:
update config

update config

Commit:
update tests

update tests

peterhashair commented:
@sgiehl sorry to bother got a question. Those two keep failing after switch to the MYSQLI adapter. Any idea why?
```
1) Piwik\Tests\Integration\ArchiveProcessingTest::testTableInsertBatch
2) Piwik\Tests\Integration\CronArchiveTest::test_output

```

Commit:
change from mariaDB to mysql

change from mariaDB to mysql

Commit:
Update matomo-tests.yml

Commit:
Update setup_mysql_tmpfs.sh

Commit:
update mysql setup

update mysql setup

Commit:
update tests

update tests

Commit:
Update setup_mysql_tmpfs.sh

Commit:
Update setup_mysql_tmpfs.sh

Commit:
Update setup_mysql_tmpfs.sh

Commit:
update password

update password

Commit:
Update matomo-tests.yml

Commit:
update to mysql

update to mysql

Commit:
update mysql config

Commit:
update config

update config

Commit:
update

update

Commit:
revert to mariadb

revert to mariadb

Commit:
Update BatchInsert.php

test without local

Commit:
Update BatchInsert.php

reset local

Commit:
Update karma.conf.js

remove plugins/Morpheus/javascripts/ajaxHelper.js

Commit:
Update BatchInsert.php

debug error

Commit:
remove var_dump

remove var_dump

Commit:
Update Comparisons.store.ts

add ts ignore

Commit:
update ts error

update ts error

Commit:
Merge remote-tracking branch 'origin/4.x-dev' into github-aciton

Commit:
Merge branch 'hotfix-random-ui-fix' into github-aciton

peterhashair committed:
built vue files

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/Diagnostics/tests/UI/expected-screenshots/Diagnostics_page.png
#	tests/UI/expected-screenshots/IncompletePeriodVisualisation_visitors_overview.png

Commit:
update all

update all

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites.png
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites_changed_sort_order.png
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites_page_1.png
#	plugins/MultiSites/tests/UI/expected-screenshots/MultiSitesTest_all_websites_search.png
#	tests/UI/expected-screenshots/UIIntegrationTest_widgetize_allwebsites.png

Commit:
Merge branch '4.x-dev' into github-aciton

Commit:
update tests

update tests

Commit:
update prepare

update prepare

Commit:
Update FrontControllerTest.php

remove all the numbers

Commit:
Update prepare.sh

update host for js

Commit:
Update matomo-tests.yml

update follow run

Commit:
Update matomo-tests.yml

update work flow

Commit:
update work flow

update work flow

Commit:
Update matomo-tests.yml

remove schedule

Commit:
update name

update name

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	tests/PHPUnit/System/expected/test_apiGetReportMetadata__API.getSegmentsMetadata.xml

Commit:
update workflow

update workflow

Commit:
Update buildvue.yml

update vue build

Commit:
Update buildvue.yml

add cancel last build

Commit:
update tests

update tests

Commit:
Update matomo-tests.yml

update on pull request

Commit:
Update piwik.js

test trigger js

Commit:
test work follow

test work follow

Commit:
update work flow

update work flow

Commit:
Merge branch '4.x-dev' into github-aciton

Commit:
Update matomo-tests.yml

update workflow

Commit:
Update matomo-tests.yml

revert node to 14

Commit:
update upload

update upload

Commit:
update tests

update tests

Commit:
update screenshots

update screenshots

Commit:
update screenshot

update screenshot

Commit:
update test

update test

Commit:
update pdf

update pdf

Commit:
Update prepare.sh

update tests

Commit:
Merge branch '4.x-dev' into github-aciton

# Conflicts:
#	plugins/CorePluginsAdmin/tests/UI/expected-screenshots/TagManagerTeaser_admin_page_disable.png
#	plugins/Dashboard/tests/UI/expected-screenshots/Dashboard_removed.png
#	plugins/Installation/tests/UI/expected-screenshots/Installation_system_check.png
#	tests/UI/expected-screenshots/UIIntegrationTest_admin_home.png
#	tests/UI/expected-screenshots/UIIntegrationTest_api_error.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard1.png
#	tests/UI/expected-screenshots/UIIntegrationTest_dashboard2.png

Commit:
Update matomo-tests.yml

try node 12

sgiehl commented:
@peterhashair Is #18679 a replacement for this one

peterhashair commented:
@sgiehl yes, this one got some extra changes on screenshot, which doesn't needed
